api/authenticator.py

```diff
@@ -232,69 +579,704 @@ def create_authentication_headers(self):
+
+        Example:
+        {
+            "name": "My OAuth Provider",
+            "methods": {
+                "http://librarysimplified.org/authtype/MyOAuthProvider" : {
+                "links": {
+                    "authenticate": "https://circulation.library.org/oauth_authenticate?provider=MyOAuth"
+                 }
+            }
+        }
+        """
         method_doc = dict(links=dict(authenticate=self._internal_authenticate_url()))
         methods = {}
         methods[self.METHOD] = method_doc
         return dict(name=self.NAME, methods=methods)
+
+    def token_data_source(self, _db):
+        # TODO: This is one of those situations where we'd like to
+        # create the data source if it doesn't already exist.
+        return DataSource.lookup(_db, self.TOKEN_DATA_SOURCE_NAME)
+
+
+class OAuthController(object):
+
+    """A controller for handling requests that are part of the OAuth
+    credential dance.
+    """
+    
+    def __init__(self, authenticator):
+        self.authenticator = authenticator
+
+    def oauth_authentication_redirect(self, params):
+        """Redirect an unauthenticated patron to the authentication URL of the
+        appropriate OAuth provider.
+
+        Over on that other site, the patron will authenticate and be
+        redirected back to the circulation manager (the URL is stored
+        in params['redirect_uri']), ending up in
+        oauth_authentication_callback.
+        """
+        redirect_uri = params.get('redirect_uri') or ""
+
+        provider_name = params.get('provider')
+        provider = self.authenticator.oauth_provider_lookup(provider_name)
+        if isinstance(provider, ProblemDetail):
+            return self._redirect_with_error(redirect_uri, provider)
+        state = dict(
+            provider=provider.NAME, redirect_uri=redirect_uri
+        )
+        state = json.dumps(state)
+        state = urllib.quote(state)
+        return redirect(provider.external_authenticate_url(state))
+
+    def oauth_authentication_callback(self, _db, params):
+        """Create a Patron object and a bearer token for a patron who has just
+        authenticated with one of our OAuth providers.
+
+        :return: A redirect to the `redirect_uri` kept in
+        `params['state']`, with the bearer token encoded into the
+        fragment identifier as `access_token` and useful information
+        about the patron encoded into the fragment identifier as
+        `patron_info`. For example, if params is 
+
+            dict(state="http://oauthprovider.org/success")
+
+        Then the redirect URI might be:
+
+            http://oauthprovider.org/success#access_token=1234&patron_info=%7B%22name%22%3A+%22Mary+Shell%22%7D
+
+        It's the client's responsibility to extract the access_token,
+        start using it as a bearer token, and make sense of the
+        patron_info.
+        """
+        code = params.get('code')
```

Comment:
Hmm, yeah, let's take another pass at this when we do a second OAuth integration.